src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -2600,5 +2623,273 @@ private void Remove(int startIndex, int count, out StringBuilder chunk, out int
+            /// <summary>Whether <see cref="_provider"/> provides an ICustomFormatter.</summary>
+            /// <remarks>
+            /// Custom formatters are very rare.  We want to support them, but it's ok if we make them more expensive
+            /// in order to make them as pay-for-play as possible.  So, we avoid adding another reference type field
+            /// to reduce the size of the builder and to reduce required zero'ing, by only storing whether the provider
+            /// provides a formatter, rather than actually storing the formatter.  This in turn means, if there is a
+            /// formatter, we pay for the extra interface call on each AppendFormatted that needs it.
+            /// </remarks>
+            private readonly bool _hasCustomFormatter;
+
+            /// <summary>Initializes the builder.</summary>
+            /// <param name="stringBuilder">The associated StringBuilder.</param>
+            private InterpolatedAppendBuilder(StringBuilder stringBuilder)
+            {
+                _stringBuilder = stringBuilder;
+                _provider = null;
+                _hasCustomFormatter = false;
+            }
+
+            /// <summary>Initializes the builder.</summary>
+            /// <param name="stringBuilder">The associated StringBuilder.</param>
+            /// <param name="provider">An object that supplies culture-specific formatting information.</param>
+            private InterpolatedAppendBuilder(StringBuilder stringBuilder, IFormatProvider? provider)
+            {
+                _stringBuilder = stringBuilder;
+                _provider = provider;
+                _hasCustomFormatter = provider is not null && InterpolatedStringBuilder.HasCustomFormatter(provider);
+            }
+
+            /// <summary>Creates a builder used to append an interpolated string into a <see cref="StringBuilder"/>.</summary>
+            /// <param name="literalLength">The number of constant characters outside of holes in the interpolated string.</param>
+            /// <param name="formattedCount">The number of holes in the interpolated string.</param>
+            /// <param name="stringBuilder">The associated StringBuilder to which to append.</param>
+            /// <remarks>This is intended to be called only by compiler-generated code. Arguments are not validated as they'd otherwise be for members intended to be used directly.</remarks>
+            public static InterpolatedAppendBuilder Create(int literalLength, int formattedCount, StringBuilder stringBuilder) =>
+                new InterpolatedAppendBuilder(stringBuilder);
+
+            /// <summary>Creates a builder used to translate an interpolated string into a <see cref="string"/>.</summary>
+            /// <param name="literalLength">The number of constant characters outside of holes in the interpolated string.</param>
+            /// <param name="formattedCount">The number of holes in the interpolated string.</param>
```

Comment:
All of the references to "holes" probably want to be "placeholders", or something from a language spec.

Reply:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated calls them "interpolation expressions", so I'll go with that.